_posts/2020-01-06-using-strimzi-with-amazon-nlb-loadbalancers.md

```diff
@@ -0,0 +1,348 @@
+Apache Kafka and Strimzi will work with this out of the box without any additional changes.
+However, it will still create a separate load balancer instance for each Kafka broker plus the additional one for the bootstrap service.
+Once the cluster is deployed, you can easily connect to it using the load balancer DNS name and the port `9094`:
+
+```
+$ bin/kafka-console-consumer.sh --bootstrap-server a83231a2e37594f3fbf5c6841de6f729-f13dac31e8214877.elb.us-east-1.amazonaws.com:9094 --topic my-topic --from-beginning
+Hello
+World
+```
+
+## Single Network Load Balancer for the whole Apache Kafka cluster
+
+But what if we want to share a single load balancer instance for all services?
+I didn't found any way how to tell Kubernetes to create only single NLB instance and share it with multiple services.
+So we will need to approach this a little differently.
+When you expose a Kubernetes service with load balancer, it essentially creates a node ports on your Kubernetes workers and configures the load balancer to forward the data to the right node port.
+So to workaround the limitations we can do the same manually.
+We can:
+* Expose Apache Kafka using node ports
+* Create NLB and configure it to route the data to the node port services
+* Configure the advertised addresses in Apache Kafka to use the NLB address instead of the node addresses and node ports
+
+Let's look in detail on each of these steps.
+
+### Exposing Apache Kafka with node ports
+
+First, we will deploy a regular Kafka cluster with node port access:
+
+```yaml
+    # ...
+    listeners:
+      # ...
+      external:
+        type: nodeport
+        tls: false
+    # ...
+```
+
+We will have to note the node port numbers of the node port services, because we will need them in the next step:
+
+```
+kubectl get service
+NAME                                  TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                      AGE
+my-cluster-kafka-0                    NodePort    10.96.84.161   <none>        9094:31208/TCP               35m
+my-cluster-kafka-1                    NodePort    10.96.63.131   <none>        9094:31983/TCP               35m
+my-cluster-kafka-2                    NodePort    10.96.35.251   <none>        9094:32279/TCP               35m
+my-cluster-kafka-bootstrap            ClusterIP   10.96.17.28    <none>        9091/TCP,9092/TCP,9093/TCP   35m
+my-cluster-kafka-brokers              ClusterIP   None           <none>        9091/TCP,9092/TCP,9093/TCP   35m
+my-cluster-kafka-external-bootstrap   NodePort    10.96.187.42   <none>        9094:32200/TCP               35m
+my-cluster-zookeeper-client           ClusterIP   10.96.58.200   <none>        2181/TCP                     36m
+my-cluster-zookeeper-nodes            ClusterIP   None           <none>        2181/TCP,2888/TCP,3888/TCP   36m
+```
+
+In my case, the node ports are:
+* `32200`
+* `31208`
+* `31983`
+* `32279`
```

Comment:
I would write a little bit differently. Something like:

* `32200` for the Kafka external bootstrap service
* `31208`, `31983` and `32279` for the related Kafka broker services

for not providing just a list of numbers.

Reply:
Or even:

* `my-cluster-kafka-0` has port `31208`
* `my-cluster-kafka-0` has port `31983`
* `my-cluster-kafka-0` has port `32279`
* `my-cluster-kafka-external-bootstrap` has port `32200`

But in any case, it's just slightly confusing that the numbers you list are in a different order to the command output.